### PR TITLE
Add correctness tests for binary operations

### PR DESCRIPTION
**Pull request by rdspring1**
This PR checks the correctness of Nvfuser binary operations against a PyTorch or Jax reference function using inputs from the `elementwise_binary_generator` function.

This PR adds tests for the following operations:
- add, atan2, bitwise_and, bitwise_left_shift, bitwise_or, bitwise_right_shift, bitwise_xor, div, eq, fmod, gcd, ge, gt, le, logical_right_shift, lt, mul, ne, nextafter, pow, remainder, sub, truediv

Remaining binary operations:
- `ceildiv` - It isn't exposed in the python frontend 
- `mod` - It requires a python-only reference since an equivalent function doesn't exist in Jax or PyTorch.

Misc:
* `div` operator is tested with `torch.div(rounding_mode='none')` for float dtypes and `torch.div(rounding_mode='trunc')` for integer dtypes.
* Rename elementwise_unary_ops to unary_ops

### CODE REVIEW

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

**Comment by kevinstephano** [comment]
Could you add a more thorough description of what this PR is trying to accomplish so I know what needs to be reviewed?

File: python_tests/pytest_input_generators.py [hunk 1]
```diff
@@ -360,6 +360,59 @@ def define_vector_input_error_generator(
...
+        device="cuda",
+        dtype=dtype,
+        low=low,
+        high=high,
+        requires_grad=requires_grad,
+        **kwargs,
+    )
+
+    shapes = (
+        # TODO: restore size zero cases
```
Comment (jjsjann123) [inline python_tests/pytest_input_generators.py:385]: is size zero broken? do we have an issue tracking it?

File: python_tests/pytest_opinfos.py [hunk 2]
```diff
@@ -350,10 +360,242 @@
...
+# atan2 --- promote int to float; allows fp16 and bf16
+# nextafter, truediv  --- promote int to float; requires full-precision fp32, fp64
+# ceildiv, div, fmod, mod, remainder, truediv --- except_zero
+# add, mul, pow, sub
+# bitwise_and, bitwise_or, bitwise_xor --- bool_int_only
+# bitwise_left_shift, bitwise_right_shift, logical_right_shift --- int_only
+# eq, ne, ge, gt, le, lt --- compare
+
+# TODO Add "ceildiv" to python_frontend
+# TODO Add support for python reference for "mod".
```
Comment (jjsjann123) [inline python_tests/pytest_opinfos.py:378]: does torch.remainder fail?

File: python_tests/pytest_utils.py [hunk 3]
```diff
@@ -19,43 +19,35 @@ class ArgumentType(Enum):
...
+
+complex_dtypes = (
     torch.complex64,
     torch.complex128,
 )
 
-int_dtypes = (torch.int32, torch.int64)
-float_dtypes = (torch.float32, torch.float64, torch.bfloat16, torch.float16)
-complex_dtypes = (torch.complex64, torch.complex128)
+# uint8, int8, int16, bf16, fp16 are skipped because nvfuser upcasts those dtypes to fp32,
```
Comment (jjsjann123) [inline python_tests/pytest_utils.py:44]: nice clean up!

File: python_tests/pytest_utils.py [hunk 4]
```diff
@@ -19,43 +19,35 @@ class ArgumentType(Enum):
...
+# but does not return the original type.
+bool_int_dtypes = bool_dtypes + int_dtypes
+int_float_dtypes = int_dtypes + float_dtypes
+float_complex_dtypes = float_dtypes + complex_dtypes
+all_dtypes_except_bool = int_dtypes + float_dtypes + complex_dtypes
+all_dtypes = bool_dtypes + all_dtypes_except_bool
```
Comment (jjsjann123) [inline python_tests/pytest_utils.py:50]: I notice that we are not putting `half_precision_float_dtypes` in any of the aggregated types above. This is surprising, it should at least in `all_dtypes_XXX` category.

**Review by jjsjann123 — COMMENTED** [review]
> mod - It requires a python-only reference since an equivalent function doesn't exist in Jax or PyTorch.

Maybe use `torch.remainder` for mod.

File: python_tests/pytest_utils.py (hunk 4 above)
Comment (rdspring1) [reply python_tests/pytest_utils.py:50]: Nvfuser promotes `bf16` and `fp16` to `fp32` but doesn't return the original dtype. For half precision types, I was thinking of skipping the dtype check.

Comment (jjsjann123) [reply python_tests/pytest_utils.py:50]: That's reasonable. But the WAR should be reflected in code.
Maybe put a comment here on why we are skipping half precision and also rename the `all_dtypes_XXX` variants to something like `all_dtypes_except_reduced`

File: python_tests/pytest_opinfos.py (hunk 2 above)
Comment (rdspring1) [reply python_tests/pytest_opinfos.py:378]: Yes.

File: python_tests/pytest_input_generators.py (hunk 1 above)
Comment (rdspring1) [reply python_tests/pytest_input_generators.py:385]: Size zero works fine. I took these test cases from elsewhere.

**Comment by kevinstephano** [comment]
I would not worry about `ceilDiv` at this time.  I think it is an operation that is more important to codegen for blocking of kernels than is needed for frameworks.

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
I was planning to add the `mod` reference in a separate PR.

**Review by jjsjann123 — APPROVED** [review]
LGTM thanks for the clean up

**Comment by rdspring1** [comment]
!build

**Pull request merged**